tests/generators/random/generate.py

```diff
@@ -0,0 +1,245 @@
+)
+from eth2spec.test.helpers.constants import PHASE0, ALTAIR
+
+
+# Ensure this many blocks are present in *each* randomized scenario
+BLOCK_TRANSITIONS_COUNT = 2
+
+
+def _normalize_transition(transition):
+    """
+    Provide "empty" or "no op" sub-transitions
+    to a given transition.
+    """
+    if isinstance(transition, Callable):
+        transition = transition()
+    if "epochs_to_skip" not in transition:
+        transition["epochs_to_skip"] = 0
+    if "slots_to_skip" not in transition:
+        transition["slots_to_skip"] = 0
+    if "block_producer" not in transition:
+        transition["block_producer"] = no_block
+    if "validation" not in transition:
+        transition["validation"] = no_op_validation
+    return transition
+
+
+def _normalize_scenarios(scenarios):
+    """
+    "Normalize" a "scenario" so that a producer of a test case
+    does not need to provide every expected key/value.
+    """
+    for scenario in scenarios:
+        transitions = scenario["transitions"]
+        for i, transition in enumerate(transitions):
+            transitions[i] = _normalize_transition(transition)
+
+
+def _flatten(t):
+    leak_transition = t[0]
+    result = [leak_transition]
+    for transition_batch in t[1]:
+        for transition in transition_batch:
+            result.append(transition)
+    return result
+
+
+def _generate_randomized_scenarios(block_randomizer):
+    """
+    Generates a set of randomized testing scenarios.
+    Return a sequence of "scenarios" where each scenario:
+    1. Provides some setup
+    2. Provides a sequence of transitions that mutate the state in some way,
+        possibly yielding blocks along the way
+    NOTE: scenarios are "normalized" with empty/no-op elements before returning
+    to the test generation to facilitate brevity when writing scenarios by hand.
+    NOTE: the main block driver builds a block for the **next** slot, so
+    the slot transitions are offset by -1 to target certain boundaries.
+    """
+    # go forward 0 or 1 epochs
+    epochs_set = (
+        epoch_transition(n=0),
+        epoch_transition(n=1),
+    )
+    # within those epochs, go forward to:
+    slots_set = (
+        # the first slot in an epoch (see note in docstring about offsets...)
+        slot_transition(last_slot_in_epoch),
+        # the second slot in an epoch
+        slot_transition(n=0),
+        # some random number of slots, but not at epoch boundaries
+        slot_transition(random_slot_in_epoch),
+        # the last slot in an epoch (see note in docstring about offsets...)
+        slot_transition(penultimate_slot_in_epoch),
+    )
+    # and produce a block...
+    blocks_set = (
+        transition_with_random_block(block_randomizer),
+    )
+    # build a set of block transitions from combinations of sub-transitions
+    transitions_generator = (
+        itertools.product(epochs_set, slots_set, blocks_set) for
+        _ in range(BLOCK_TRANSITIONS_COUNT)
```

Comment:
The combinatorics didn't work as I immediately expected them too.

We always couple like-randomness with each other (e.g. two sets of (no epoch, go to pen-ultimate slot, random block)) rather than mixing the different pre-block modifications (e.g. (no epoch, penultimate slot, random block) + (1 epoch, random slot, random block)  )

Can you explain your thinking about here? On the one hand, we could do many more combinations but the number would explode. Also, I suppose that the magic is primarily in fussing with slots/epoch and then random blocks rather than fussing with slots/epochs in different ways between random blocks..